File: gtran-api/app/domain/mission_profit.py
from __future__ import annotations

from app.schemas import FuelRecord, Mission


def compute_mission_profit(mission: Mission, fuel_records: list[FuelRecord]) -> dict:
    vehicle_fuel = sum(f.montant for f in fuel_records if f.vehicleId == mission.vehicleId)
    carburant = min(vehicle_fuel, mission.cout * 0.5) if vehicle_fuel > 0 else round(mission.cout * 0.36)
    peages = round(mission.cout * 0.12)
    salaire = round(mission.cout * 0.28)
    reparations = round(mission.cout * 0.08)
    autres = max(0, mission.cout - carburant - peages - salaire - reparations)
    benefice = mission.prix - mission.cout

    return {
        "prix": mission.prix,
        "carburant": carburant,
        "peages": peages,
        "salaire": salaire,
        "reparations": reparations,
        "autres": autres,
        "coutTotal": mission.cout,
        "benefice": benefice,
        "margePct": (benefice / mission.prix * 100) if mission.prix > 0 else 0,
    }


def format_mission_profit_text(mission: Mission, breakdown: dict) -> str:
    fmt = lambda n: f"{int(n):,}".replace(",", " ")
    return "\n".join(
        [
            f"Mission {mission.id} ({mission.client}, {mission.depart} → {mission.destination}) :",
            f"• Prix de vente : {fmt(breakdown['prix'])} XOF",
            f"• Carburant : {fmt(breakdown['carburant'])} XOF",
            f"• Péages : {fmt(breakdown['peages'])} XOF",
            f"• Salaire chauffeur : {fmt(breakdown['salaire'])} XOF",
            f"• Réparations / divers : {fmt(breakdown['reparations'] + breakdown['autres'])} XOF",
            f"• Coût total : {fmt(breakdown['coutTotal'])} XOF",
            f"• Bénéfice net : {fmt(breakdown['benefice'])} XOF ({breakdown['margePct']:.1f}% de marge)",
        ]
    )

File: gtran-app/src/lib/mission-profit.ts
import type { FuelRecord, Mission } from "@/types/entities"

export interface MissionProfitBreakdown {
  prix: number
  carburant: number
  peages: number
  salaire: number
  reparations: number
  autres: number
  coutTotal: number
  benefice: number
  margePct: number
}

export function computeMissionProfit(mission: Mission, fuelRecords: FuelRecord[]): MissionProfitBreakdown {
  const vehicleFuel = fuelRecords
    .filter((f) => f.vehicleId === mission.vehicleId)
    .reduce((s, f) => s + f.montant, 0)

  const carburant = vehicleFuel > 0 ? Math.min(vehicleFuel, mission.cout * 0.5) : Math.round(mission.cout * 0.36)
  const peages = Math.round(mission.cout * 0.12)
  const salaire = Math.round(mission.cout * 0.28)
  const reparations = Math.round(mission.cout * 0.08)
  const autres = Math.max(0, mission.cout - carburant - peages - salaire - reparations)
  const benefice = mission.prix - mission.cout

  return {
    prix: mission.prix,
    carburant,
    peages,
    salaire,
    reparations,
    autres,
    coutTotal: mission.cout,
    benefice,
    margePct: mission.prix > 0 ? (benefice / mission.prix) * 100 : 0,
  }
}

export function formatMissionProfitText(mission: Mission, breakdown: MissionProfitBreakdown): string {
  const fmt = (n: number) => n.toLocaleString("fr-FR")
  return [
    `Mission ${mission.id} (${mission.client}, ${mission.depart} → ${mission.destination}) :`,
    `• Prix de vente : ${fmt(breakdown.prix)} XOF`,
    `• Carburant : ${fmt(breakdown.carburant)} XOF`,
    `• Péages : ${fmt(breakdown.peages)} XOF`,
    `• Salaire chauffeur : ${fmt(breakdown.salaire)} XOF`,
    `• Réparations / divers : ${fmt(breakdown.reparations + breakdown.autres)} XOF`,
    `• Coût total : ${fmt(breakdown.coutTotal)} XOF`,
    `• Bénéfice net : ${fmt(breakdown.benefice)} XOF (${breakdown.margePct.toFixed(1)}% de marge)`,
  ].join("\n")
}
